dev/crypto.md

```diff
@@ -91,29 +89,144 @@ for the Byzantine Fault Tolerance protocol, and (2) rerandomize its
+    return [0 for _ in range(32)]
+  if len(hashes) == 1:
+    return hashes[0]
+  nexthashes = []
+  while len(hashes) > 0:
+    left = hashes[0]
+    right = hashes[1] if len(hashes) > 1 else [0 for _ in range(32)]
+    hashes = hashes[2:]
+    nexthashes.append(H("MA" + left + right))
+  return reduce(nexthashes)
+```
+
+## Proofs
+
+Logically, to verify that an element appears at some position P in the
+array, the verifier runs a variant of the commit procedure to compute
+a candidate root hash, and then checks if the resulting root hash is
+equal to the expected commitment value.  The key difference is that the
+verifier does not have access to the entire list of committed elements;
+the verifier just has some subset of elemens (one or more), along with the
+positions at which these elements appear.  Thus, the verifier needs to
+know the siblings (the `left` and `right` values used in the `reduce()`
+function above) to compute its candidate root hash.  The list of these
+siblings constitutes the proof; thus, a proof is a list of zero or more
+32-byte hash values.  Algorand defines a deterministic order in which
+the verification procedure expects to find siblings in this proof, so no
+additional information is required as part of the proof (in particular,
+no information about which part of the Merkle tree each proof element
+corresponds to).
+
+## Verifying a proof
+
+The following pseudocode defines the logic for verifying a proof (a
+list of 32-byte hashes) for one or more elements, specified as a list
+of position-element pairs, sorted by position in the array, against a
+root commitment.  The function `verify` returns `True` if `proof` is a
+valid proof for all elements in `elems` being present at their positions
+in the array committed to by `root`.  (The pseudocode might raise an
+exception due to accessing the proof past the end; this is equivalent
+to returning `False`.)  The function implements a variant of `reduce()`
+for a sparse array, rather than a fully-populated one.
+
+```
+def verify(elems, proof, root):
+  if len(elems) == 0:
+    return len(proof) == 0
+  if len(elems) == 1 and len(proof) == 0:
+    return elems[0].pos == 0 && elems[0].hash == root
+
+  i = 0
+  nextelems = []
+  while i < len(elems):
+    pos = elems[i].pos
+    poshash = elems[i].hash
+    sibling = pos ^ 1
+    if i+1 < len(elems) and elems[i+1].pos == sibling:
+      sibhash = elems[i+1].hash
+      i++
+    else:
+      sibhash = proof[0]
+      proof = proof[1:]
+    if pos&1 == 0:
+      h = H("MA" + poshash + sibhash)
+    else:
+      h = H("MA" + sibhash + poshash)
+    nextelems.append({"pos": pos/2, "hash": h})
+
+  return verify(nextelems, proof, root)
+```
 
 
 
-Verifiable Random Function
-==========================
-
-
-Cryptographic Sortition
------------------------
-
-
 [ledger-spec]: https://github.com/algorand/spec/ledger.md
 [abft-spec]: https://github.com/algorand/spec/abft.md
 
 [sha]: TODO-sha
 [ed25519]: TODO-ed25519
```

Comment:
Should https://tools.ietf.org/html/rfc8032 be used as reference for ed25519?

Reply:
Sure, might as well plug it in; thanks.